File: 1032.leetcode.py
"""
1032. Stream of Characters
https://leetcode.com/problems/stream-of-characters/
LTE
"""
class TrieNode:
    def __init__(self):
        self.is_word = False
        self.children = {}
        self.word = None
        self.prefix = set()
        
class StreamChecker:

    def __init__(self, words: List[str]):
        self.trie = TrieNode()
        self.q = deque()
        for word in words:
            root = self.trie
            for c in word:
                root.children[c] = root.children.get(c, TrieNode())
                root = root.children[c]
                root.prefix.add(c)
            root.is_word = True
            root.word = word
    
    def query(self, letter: str) -> bool:
        if letter in self.trie.children:
            self.q.append(self.trie)
        
        n = len(self.q)
        ok = False
        for _ in range(n):
            top = self.q.popleft()
            if letter in top.children:
                top = top.children[letter]
                if top.is_word:
                    ok = True
                self.q.append(top)
        
        return ok
        


# Your StreamChecker object will be instantiated and called as such:
# obj = StreamChecker(words)
# param_1 = obj.query(letter)

File: 1032.leetcode.cpp
/**
1032. Stream of Characters
https://leetcode.com/problems/stream-of-characters/
reverse char -> trie
**/
class TrieNode {
public:
    map<char, TrieNode* > children;
    bool is_word = false;
};
class StreamChecker {
private:
    TrieNode * const root = new TrieNode();
    deque<char> stream;
public:
    StreamChecker(vector<string>& words) {
        for (string& w : words) {
            TrieNode *node = root;
            reverse(w.begin(), w.end());
            for (char& c: w) {
                if (node->children.count(c) == 0) node->children[c] = new TrieNode();
                node = node->children[c];
            }
            node->is_word = true;
        }
    }
    
    bool query(char c) {
        stream.push_front(c);
        TrieNode *node = root;
        for (char& c: stream) {
            if (node->children.count(c) == 0) return false;
            node = node->children[c];
            if (node->is_word) return true;
        }
        return false;
    }
};